### PR TITLE
TypeScript: Support type arguments on tagged templates

### PR DESCRIPTION
**Pull request by andy-ms**
| Q                        | A
| ------------------------ | ---
| Fixed Issues?            | #7747 (partly)
| Patch: Bug Fix?          | 
| Major: Breaking Change?  | 
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR         |
| Any Dependency Changes?  |
| License                  | MIT

@JamesHenry This changes the AST format. CC @DanielRosenwasser for review.
Supports parsing type arguments on tagged template calls.
Should wait on Microsoft/TypeScript#23430 to be merged so we're sure we have the final syntax.

### CODE REVIEW

**Comment by existentialism** [comment]
LGTM, but will wait for @JamesHenry/@DanielRosenwasser to chime in!

**Comment by DanielRosenwasser** [comment]
I feel like I commented this before, but I guess not. You need a test like

```ts
new foo<A, B>
`hello world`
```

To show ASI doesn't kick in.

**Comment by DanielRosenwasser** [comment]
I feel like I commented this before, but I guess not. You need a test like

```ts
new foo<A, B>
`hello world`
```

To show ASI doesn't kick in.
I feel like I commented this before, but I guess not. You need a test like

```ts
new foo<A, B>
`hello world`
```

To show ASI doesn't kick in.
I feel like I commented this before, but I guess not. You need a test like

```ts
new foo<A, B>
`hello world`
```

To show ASI doesn't kick in.
I feel like I commented this before, but I guess not. You need a test like

```ts
new foo<A, B>
`hello world`
```

To show ASI doesn't kick in.

**Comment by DanielRosenwasser** [comment]
Christ, mobile GitHub is a real piece of work.

**Comment by existentialism** [comment]
@DanielRosenwasser well, at least you made your point :)

File: packages/babylon/test/fixtures/typescript/type-arguments/tagged-template-no-asi/output.json [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+              "line": 1,
+              "column": 0
+            },
+            "end": {
+              "line": 2,
+              "column": 2
+            }
+          },
+          "tag": {
+            "type": "NewExpression",
```
Comment (DanielRosenwasser) [inline packages/babylon/test/fixtures/typescript/type-arguments/tagged-template-no-asi/output.json]: This looks wrong. The tagged template should have higher precedence than the *NewExpression*. As an example, in the following:

```js
new C
``
```

here's what I currently get:

```json5
{
  // ...
  "expression": {
    "type": "NewExpression",
    "start": 0,
    "end": 8,
    "loc": {
      "start": {
        "line": 1,
        "column": 0
      },
      "end": {
        "line": 2,
        "column": 2
      }
    },
    "callee": {
      "type": "TaggedTemplateExpression",
      "start": 4,
      "end": 8,
      "loc": {
        "start": {
          "line": 1,
          "column": 4
        },
        "end": {
          "line": 2,
          "column": 2
        }
      },
      "tag": {
        "type": "Identifier",
        "start": 4,
        "end": 5,
        "loc": {
          "start": {
            "line": 1,
            "column": 4
          },
          "end": {
            "line": 1,
            "column": 5
          },
          "identifierName": "C"
        },
        "name": "C"
      },
      "quasi": {
        "type": "TemplateLiteral",
        "start": 6,
        "end": 8,
        "loc": {
          "start": {
            "line": 2,
            "column": 0
          },
          "end": {
            "line": 2,
            "column": 2
          }
        },
        "expressions": [],
        "quasis": [
          {
            "type": "TemplateElement",
            "start": 7,
            "end": 7,
            "loc": {
              "start": {
                "line": 2,
                "column": 1
              },
              "end": {
                "line": 2,
                "column": 1
              }
            },
            "value": {
              "raw": "",
              "cooked": ""
            },
            "tail": true
          }
        ]
      }
    },
    "arguments": []
  }
}
```

**Comment by kachkaev** [comment]
Given that styled-components accept types in template literals (https://github.com/styled-components/styled-components/pull/1798), lot's of people would love to see this merged!

TypeScript is transpiled with babel in Next.js, which is pretty popular and is often used with styled-components. Here is a downstream issue that's pending this PR: https://github.com/zeit/next-plugins/issues/203. 🙏

**Comment by EItanya** [comment]
Any update on this being merged, Would love to be able to start using this feature!

**Comment by timche** [comment]
Is there anything blocking from merging this?

**Pull request merged**